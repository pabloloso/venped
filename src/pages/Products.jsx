import React, { useEffect, useRef, useState } from 'react';
import { gql, useLazyQuery, useQuery } from '@apollo/client';

import Sidebar from '../components/Sidebar';
import Header from '../components/Header';
import SearchForm from '../components/SearchForm';
import DropdownFilter from '../components/DropdownFilter';
import ProductsTable from '../components/ProductsTable';
import PaginationNumeric from '../components/PaginationNumeric';

const query = gql`
  query {
    fetchProducts {
      results(
        page: 1,
        perPage: 10
      ) {
        id,
        title,
        price,
        tax,
        stock
      },
      pagination(
        page: 1,
        perPage: 10
      ) {
        totalResults,
        limitValue,
        totalPages,
        currentPage,
        nextPage,
        prevPage,
        firstPage,
        lastPage,
        outOfRange
      }
    }
  }
`;

const prueba = gql`
  query FetchProducts (
    $tax_filter: [String!],
    $title_filter: String,
    $order_by: String,
    $order: String,
    $page: Int!,
    $per_page: Int!
  ) {
    fetchProducts {
      results(
        taxFilter: $tax_filter,
        titleFilter: $title_filter,
        orderBy: $order_by,
        order: $order,
        page: $page,
        perPage: $per_page
      ) {
        id,
        title,
        price,
        tax,
        stock
      },
      pagination(
        taxFilter: $tax_filter,
        titleFilter: $title_filter,
        orderBy: $order_by,
        order: $order,
        page: $page,
        perPage: $per_page
      ) {
        totalResults,
        limitValue,
        totalPages,
        currentPage,
        nextPage,
        prevPage,
        firstPage,
        lastPage,
        outOfRange
      }
    }
  }
`;

function Products() {
  const { loading, data } = useQuery(query);
  const [getProducts, result] = useLazyQuery(prueba);
  const { loading: loadingResult } = result;

  const [page, setPage] = useState(1);
  const [filters, setFilters] = useState([]);
  const [searchCriteria, setSearchCriteria] = useState('');
  const [orderBy, setOrderBy] = useState('');
  const [order, setOrder] = useState('');

  const [sidebarOpen, setSidebarOpen] = useState(false);

  const handleChangeCurrentPage = (currentPage) => {
    setPage(currentPage);
  };

  const handleOrderBy = (event) => {
    setOrderBy(event.target.value);
    setOrder(order === 'ASC' ? 'DESC' : 'ASC');
  };

  const resetCriterias = () => {
    setOrderBy('');
    setOrder('');
    setPage(1);
  };

  const updateFiltersRequest = (selectedFilters) => {
    setFilters([...selectedFilters]);
    resetCriterias();
  };

  const searchProduct = (searchText) => {
    setSearchCriteria(searchText);
    resetCriterias();
  };

  const firstUpdate = useRef(true);
  useEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false;
      return;
    }

    const newRequest = {
      ...filters.length > 0 && { tax_filter: filters },
      ...searchCriteria !== '' && { title_filter: searchCriteria },
      ...orderBy !== '' && { order_by: orderBy },
      ...order !== '' && { order },
      page,
      per_page: 10,
    };

    getProducts({ variables: newRequest });
  }, [filters, order, orderBy, page, searchCriteria]);

  if (loading || loadingResult) return 'Loading ....';

  return (
    <div className="flex h-screen overflow-hidden">
      <Sidebar
        sidebarOpen={sidebarOpen}
        setSidebarOpen={setSidebarOpen}
      />
      <div className="relative flex flex-col flex-1 overflow-y-auto overflow-x-hidden">
        <Header
          sidebarOpen={sidebarOpen}
          setSidebarOpen={setSidebarOpen}
        />
        <main>
          <div className="px-4 sm:px-6 lg:px-8 py-8 w-full max-w-9xl mx-auto">
            <div className="sm:flex sm:justify-between sm:items-center mb-5">
              <div className="mb-4 sm:mb-0">
                <h1 className="text-2xl md:text-3xl text-slate-800 font-bold">Cat√°logo</h1>
              </div>
              <div className="grid grid-flow-col sm:auto-cols-max justify-start sm:justify-end gap-2">
                <SearchForm
                  searchProduct={searchProduct}
                />
                <DropdownFilter
                  align="right"
                  filters={filters}
                  updateFiltersRequest={updateFiltersRequest}
                />
              </div>
            </div>
            <ProductsTable
              pagination={result.data?.fetchProducts?.pagination || data.fetchProducts.pagination}
              products={result.data?.fetchProducts?.results || data.fetchProducts.results}
              handleOrderBy={handleOrderBy}
            />
            <div className="mt-8">
              <PaginationNumeric
                pagination={result.data?.fetchProducts?.pagination || data.fetchProducts.pagination}
                handleChangeCurrentPage={handleChangeCurrentPage}
              />
            </div>
          </div>
        </main>
      </div>
    </div>
  );
}

export default Products;
