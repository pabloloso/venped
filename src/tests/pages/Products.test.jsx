import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import {
  describe,
  expect,
  test,
} from 'vitest';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';

import { getProductsQueryMock } from '../mocks/productsRequest';

import Products from '../../pages/Products';

describe('Products page', () => {
  test('should render page with graphql integration', async () => {
    const { container } = render(
      <MockedProvider mocks={getProductsQueryMock} addTypename={false}>
        <Router>
          <Products />
        </Router>
      </MockedProvider>,
    );

    // eslint-disable-next-line no-promise-executor-return
    await new Promise((resolve) => setTimeout(resolve, 0));

    expect(container).toMatchSnapshot();
  });
});
