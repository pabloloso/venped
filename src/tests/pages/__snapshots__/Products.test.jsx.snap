// Vitest Snapshot v1

exports[`Products page > should render page with graphql integration 1`] = `
<div>
  <div
    class="flex h-screen overflow-hidden"
  >
    <div>
      <div
        aria-hidden="true"
        class="
          fixed
          inset-0
          bg-slate-900
          bg-opacity-30
          z-40
          lg:hidden
          lg:z-auto
          transition-opacity
          duration-200
          opacity-0 pointer-events-none
        "
      />
      <div
        class="flex flex-col absolute z-40 left-0 top-0 lg:static lg:left-auto lg:top-auto lg:translate-x-0 transform h-screen overflow-y-scroll lg:overflow-y-auto no-scrollbar w-64 lg:w-20 lg:sidebar-expanded:!w-64 2xl:!w-64 shrink-0 bg-slate-800 p-4 transition-all duration-200 ease-in-out -translate-x-64"
        id="sidebar"
      >
        <div
          class="flex justify-between mb-10 pr-3 sm:px-2"
        >
          <button
            aria-controls="sidebar"
            aria-expanded="false"
            class="lg:hidden text-slate-500 hover:text-slate-400"
            type="button"
          >
            <span
              class="sr-only"
            >
              Close sidebar
            </span>
            <svg
              class="w-6 h-6 fill-current"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.7 18.7l1.4-1.4L7.8 13H20v-2H7.8l4.3-4.3-1.4-1.4L4 12z"
              />
            </svg>
          </button>
          <a
            aria-current="page"
            class="block active"
            href="/"
          >
            Venped
          </a>
        </div>
        <div
          class="space-y-8"
        >
          <div>
            <h3
              class="text-xs uppercase text-slate-500 font-semibold pl-3"
            >
              <span
                aria-hidden="true"
                class="hidden lg:block lg:sidebar-expanded:hidden 2xl:hidden text-center w-6"
              >
                •••
              </span>
              <span
                class="lg:hidden lg:sidebar-expanded:block 2xl:block"
              >
                pages
              </span>
            </h3>
            <ul
              class="mt-3"
            >
              <li
                class="px-3 py-2 rounded-sm mb-0.5 last:mb-0 bg-slate-900"
              >
                <a
                  class="block text-slate-200 hover:text-white truncate transition duration-150 hover:text-slate-200"
                  href="/products"
                >
                  <div
                    class="flex items-center"
                  >
                    <svg
                      class="shrink-0 h-6 w-6"
                      viewBox="0 0 24 24"
                    >
                      <path
                        class="fill-current text-slate-400 !text-blue-500"
                        d="M12 0C5.383 0 0 5.383 0 12s5.383 12 12 12 12-5.383 12-12S18.617 0 12 0z"
                      />
                      <path
                        class="fill-current text-slate-600 text-blue-600"
                        d="M12 3c-4.963 0-9 4.037-9 9s4.037 9 9 9 9-4.037 9-9-4.037-9-9-9z"
                      />
                      <path
                        class="fill-current text-slate-400 text-blue-200"
                        d="M12 15c-1.654 0-3-1.346-3-3 0-.462.113-.894.3-1.285L6 6l4.714 3.301A2.973 2.973 0 0112 9c1.654 0 3 1.346 3 3s-1.346 3-3 3z"
                      />
                    </svg>
                    <span
                      class="text-sm font-medium ml-3 lg:opacity-0 lg:sidebar-expanded:opacity-100 2xl:opacity-100 duration-200"
                    >
                      products
                    </span>
                  </div>
                </a>
              </li>
            </ul>
          </div>
        </div>
        <div
          class="pt-3 hidden lg:inline-flex 2xl:hidden justify-end mt-auto"
        >
          <div
            class="px-3 py-2"
          >
            <button
              type="button"
            >
              <span
                class="sr-only"
              >
                Expand / collapse sidebar
              </span>
              <svg
                class="w-6 h-6 fill-current sidebar-expanded:rotate-180"
                viewBox="0 0 24 24"
              >
                <path
                  class="text-slate-400"
                  d="M19.586 11l-5-5L16 4.586 23.414 12 16 19.414 14.586 18l5-5H7v-2z"
                />
                <path
                  class="text-slate-600"
                  d="M3 23H1V1h2z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="relative flex flex-col flex-1 overflow-y-auto overflow-x-hidden"
    >
      <header
        class="sticky top-0 bg-white border-b border-slate-200 z-30"
      >
        <div
          class="px-4 sm:px-6 lg:px-8"
        >
          <div
            class="flex items-center justify-between h-16 -mb-px"
          >
            <div
              class="flex"
            />
            <div
              class="flex items-center space-x-3"
            >
              <select
                class="form-select focus:border-slate-300"
              >
                <option
                  selected=""
                  value="en"
                >
                  EN
                </option>
                <option
                  value="es"
                >
                  ES
                </option>
              </select>
            </div>
          </div>
        </div>
      </header>
      <main>
        <div
          class="px-4 sm:px-6 lg:px-8 py-8 w-full max-w-9xl mx-auto"
        >
          <div
            class="sm:flex sm:justify-between sm:items-center mb-5"
          >
            <div
              class="mb-4 sm:mb-0"
            >
              <h1
                class="text-2xl md:text-3xl text-slate-800 font-bold"
              >
                products
              </h1>
            </div>
            <div
              class="grid grid-flow-col sm:auto-cols-max justify-start sm:justify-end gap-2"
            >
              <form
                class="relative"
              >
                <label
                  class="sr-only"
                  for="action-search"
                >
                  search
                </label>
                <input
                  class="form-input pl-9 focus:border-slate-300"
                  id="action-search"
                  placeholder="search"
                  type="search"
                  value=""
                />
                <button
                  aria-label="search"
                  class="absolute inset-0 right-auto group"
                  type="submit"
                >
                  <svg
                    class="w-4 h-4 shrink-0 fill-current text-slate-400 group-hover:text-slate-500 ml-3 mr-2"
                    viewBox="0 0 16 16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7 14c-3.86 0-7-3.14-7-7s3.14-7 7-7 7 3.14 7 7-3.14 7-7 7zM7 2C4.243 2 2 4.243 2 7s2.243 5 5 5 5-2.243 5-5-2.243-5-5-5z"
                    />
                    <path
                      d="M15.707 14.293L13.314 11.9a8.019 8.019 0 01-1.414 1.414l2.393 2.393a.997.997 0 001.414 0 .999.999 0 000-1.414z"
                    />
                  </svg>
                </button>
              </form>
              <div
                class="relative inline-flex"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="btn bg-white border-slate-200 hover:border-slate-300 text-slate-500 hover:text-slate-600"
                  type="button"
                >
                  <span
                    class="sr-only"
                  >
                    filter
                  </span>
                  <wbr />
                  taxes
                </button>
                <div
                  class="
          origin-top-right
          z-10
          absolute
          top-full
          min-w-56
          bg-white
          border
          border-slate-200
          pt-1.5
          rounded
          shadow-lg
          overflow-hidden
          mt-1
          right-0
        "
                  style="display: none;"
                >
                  <div>
                    <div
                      class="text-xs font-semibold text-slate-400 uppercase pt-1.5 pb-2 px-4"
                    >
                      filters
                    </div>
                    <ul
                      class="mb-4"
                    >
                      <li
                        class="py-1 px-3"
                      >
                        <label
                          class="flex items-center"
                        >
                          <input
                            class="form-checkbox"
                            type="checkbox"
                            value="es_general_21"
                          />
                          <span
                            class="text-sm font-medium ml-2"
                          >
                            es_general_21
                          </span>
                        </label>
                      </li>
                      <li
                        class="py-1 px-3"
                      >
                        <label
                          class="flex items-center"
                        >
                          <input
                            class="form-checkbox"
                            type="checkbox"
                            value="es_reduced_10"
                          />
                          <span
                            class="text-sm font-medium ml-2"
                          >
                            es_general_10
                          </span>
                        </label>
                      </li>
                      <li
                        class="py-1 px-3"
                      >
                        <label
                          class="flex items-center"
                        >
                          <input
                            class="form-checkbox"
                            type="checkbox"
                            value="es_super-reduced_4"
                          />
                          <span
                            class="text-sm font-medium ml-2"
                          >
                            es_super-reduced_4
                          </span>
                        </label>
                      </li>
                      <li
                        class="py-1 px-3"
                      >
                        <label
                          class="flex items-center"
                        >
                          <input
                            class="form-checkbox"
                            type="checkbox"
                            value="fr_general_20"
                          />
                          <span
                            class="text-sm font-medium ml-2"
                          >
                            fr_general_20
                          </span>
                        </label>
                      </li>
                      <li
                        class="py-1 px-3"
                      >
                        <label
                          class="flex items-center"
                        >
                          <input
                            class="form-checkbox"
                            type="checkbox"
                            value="fr_reduced_5.5"
                          />
                          <span
                            class="text-sm font-medium ml-2"
                          >
                            fr_reduced_5.5
                          </span>
                        </label>
                      </li>
                    </ul>
                    <div
                      class="py-2 px-3 border-t border-slate-200 bg-slate-50"
                    >
                      <ul
                        class="flex items-center justify-between"
                      >
                        <li>
                          <button
                            class="btn-xs bg-white border-slate-200 hover:border-slate-300 text-slate-500 hover:text-slate-600"
                            type="button"
                          >
                            clear
                          </button>
                        </li>
                        <li>
                          <button
                            class="btn-xs bg-blue-500 hover:bg-blue-600 text-white"
                            type="button"
                          >
                            apply
                          </button>
                        </li>
                      </ul>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="bg-white shadow-lg rounded-sm border border-slate-200 relative"
          >
            <header
              class="px-5 py-4"
            >
              <h2
                class="font-semibold text-slate-800"
              >
                products
                 
                <span
                  class="text-slate-400 font-medium"
                >
                  953
                </span>
              </h2>
            </header>
            <div>
              <div
                class="overflow-x-auto"
              >
                <table
                  class="table-auto w-full"
                >
                  <thead
                    class="text-xs font-semibold uppercase text-slate-500 bg-slate-50 border-t border-b border-slate-200"
                  >
                    <tr>
                      <th
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <button
                          class="w-full font-semibold text-left"
                          type="button"
                          value="id"
                        >
                          SKU
                        </button>
                      </th>
                      <th
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <button
                          class="w-full font-semibold text-left"
                          type="button"
                          value="title"
                        >
                          article
                        </button>
                      </th>
                      <th
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <button
                          class="w-full font-semibold text-left"
                          type="button"
                          value="price"
                        >
                          price
                        </button>
                      </th>
                      <th
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <button
                          class="w-full font-semibold text-left"
                          type="button"
                          value="tax"
                        >
                          tax
                        </button>
                      </th>
                      <th
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <button
                          class="w-full font-semibold text-left"
                          type="button"
                          value="stock"
                        >
                          stock
                        </button>
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="text-sm divide-y divide-slate-200"
                  >
                    <tr>
                      <td
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <div
                          class="font-medium text-sky-500"
                        >
                          2
                        </div>
                      </td>
                      <td
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <div
                          class="font-medium text-gray-500"
                        >
                          Lightweight Wooden Coat
                        </div>
                      </td>
                      <td
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <div
                          class="font-medium text-gray-500"
                        >
                          69.43
                        </div>
                      </td>
                      <td
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <div
                          class="rounded-full px-3 py-1 font-medium w-fit text-violet-600 bg-violet-200"
                        >
                          es_general_21
                        </div>
                      </td>
                      <td
                        class="px-2 first:pl-5 last:pr-5 py-3 whitespace-nowrap"
                      >
                        <div
                          class="
          font-medium
          rounded-full
          w-fit
          px-3
          py-1
          text-green-600 bg-green-200
        "
                        >
                          21
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
          <div
            class="mt-8"
          >
            <div
              class="flex justify-center"
            >
              <nav
                aria-label="Navigation"
                class="flex"
                role="navigation"
              >
                <div
                  class="mr-2"
                >
                  <button
                    class="
                inline-flex
                items-center
                justify-center
                rounded
                leading-5
                px-2.5
                py-2
                bg-white
                border
                border-slate-200
                text-slate-600
                shadow-sm
                false
                cursor-not-allowed text-slate-300
              "
                    disabled=""
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      previous
                    </span>
                    <wbr />
                    <svg
                      class="h-4 w-4 fill-current"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M9.4 13.4l1.4-1.4-4-4 4-4-1.4-1.4L4 8z"
                      />
                    </svg>
                  </button>
                </div>
                <ul
                  class="inline-flex text-sm font-medium -space-x-px shadow-sm"
                >
                  <button
                    type="button"
                  >
                    <span
                      class="
                  inline-flex
                  items-center
                  justify-center
                  leading-5
                  px-3.5
                  py-2
                  bg-white
                  hover:bg-blue-500
                  border
                  border-slate-200
                  text-slate-600
                  hover:text-white bg-blue-500 text-white
                "
                    >
                      1
                    </span>
                  </button>
                  <button
                    type="button"
                  >
                    <span
                      class="inline-flex items-center justify-center leading-5 px-3.5 py-2 bg-white hover:bg-blue-500 border border-slate-200 text-slate-600 hover:text-white"
                    >
                      2
                    </span>
                  </button>
                  <li>
                    <span
                      class="inline-flex items-center justify-center leading-5 px-3.5 py-2 bg-white border border-slate-200 text-slate-600"
                    >
                      …
                    </span>
                  </li>
                  <button
                    type="button"
                  >
                    <span
                      class="
                  inline-flex
                  items-center
                  justify-center
                  leading-5
                  px-3.5
                  py-2
                  bg-white
                  hover:bg-blue-500
                  border
                  border-slate-200
                  text-slate-600
                  hover:text-white false
                "
                    >
                      96
                    </span>
                  </button>
                </ul>
                <div
                  class="ml-2"
                >
                  <button
                    class="
                inline-flex
                items-center
                justify-center
                rounded
                leading-5
                px-2.5
                py-2
                bg-white
                border
                border-slate-200
                text-slate-600
                shadow-sm
                hover:bg-blue-500 hover:text-white
                false
              "
                    type="button"
                  >
                    <span
                      class="sr-only"
                    >
                      next
                    </span>
                    <wbr />
                    <svg
                      class="h-4 w-4 fill-current"
                      viewBox="0 0 16 16"
                    >
                      <path
                        d="M6.6 13.4L5.2 12l4-4-4-4 1.4-1.4L12 8z"
                      />
                    </svg>
                  </button>
                </div>
              </nav>
            </div>
            <div
              class="flex justify-center"
            >
              <div
                class="text-sm text-slate-500 text-center sm:text-left py-2"
              >
                <span
                  class="font-medium text-slate-600"
                >
                  1
                </span>
                 
                to
                 
                <span
                  class="font-medium text-slate-600"
                >
                  10
                </span>
                 
                of
                 
                <span
                  class="font-medium text-slate-600"
                >
                  953
                </span>
                 
                results
              </div>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
